### PR TITLE
fix bug: dependent tasks are ignored when no need rebuild index

### PR DESCRIPTION
**Pull request by zhoney**
Happens when create index label

fixed: #211

Change-Id: I1c265950c4591620824dddcd82efa7371a7f3279

### CODE REVIEW

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java [hunk 1]
```diff
@@ -1316,16 +1318,14 @@ public void removeVertices(VertexLabel vertexLabel) {
...
-                                        vertexLabel, Query.NO_LIMIT);
-            int count = 0;
-            while (vertices.hasNext()) {
-                this.removeVertex((HugeVertex) vertices.next());
+            final int[] count = {0};
+            this.traverseVerticesByLabel(vertexLabel, vertex -> {
+                this.removeVertex(vertex);
                 // Avoid reaching tx limit
-                if (++count == this.vertexesCapacity) {
+                if (++count[0] == this.vertexesCapacity) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java]: commitIfGtSize

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java [hunk 2]
```diff
@@ -1045,30 +1045,31 @@ public void rebuildIndex(SchemaElement schema) {
...
             default:
                 assert schema.type() == HugeType.PROPERTY_KEY;
                 throw new AssertionError(String.format(
-                          "The %s can't rebuild index.", schema.type()));
+                          "The %s can't rebuild index", schema.type()));
         }
     }
 
-    public void rebuildIndex(HugeType type, Id label,
+    public void rebuildIndex(SchemaLabel label,
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java]: put in one line

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java [hunk 3]
```diff
@@ -1372,38 +1371,83 @@ public void removeEdges(EdgeLabel edgeLabel) {
...
-            query.eq(HugeKeys.LABEL, vertexLabel.id());
-            return this.queryVertices(query);
-        }
+
+        // Query.DEFAULT_CAPACITY vertices per pass
+        int pass = 0;
+        int counter;
+        do {
+            counter = 0;
+            query.offset(pass++ * Query.DEFAULT_CAPACITY);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java]: just set offset if vertexLabel.enableLabelIndex()

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java [hunk 4]
```diff
@@ -1102,13 +1103,33 @@ public void rebuildIndex(HugeType type, Id label,
              * They have different id lead to it can't compare and optimize
              */
             this.commit();
-            if (type == HugeType.VERTEX_LABEL) {
-                this.rebuildIndex(HugeType.VERTEX, label, indexLabelIds,
-                                  graphTx::queryVerticesFromBackend);
+            if (label.type() == HugeType.VERTEX_LABEL) {
+                graphTx.traverseVerticesByLabel(
+                        (VertexLabel) label,
+                        vertex -> {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java]: define Consumer<HugeElement> consumer
and let it be shared by traverseVerticesByLabel and traverseEdgesByLabel

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java [hunk 5]
```diff
@@ -1372,38 +1363,94 @@ public void removeEdges(EdgeLabel edgeLabel) {
...
         if (!edgeLabel.enableLabelIndex()) {
-            return new FilterIterator<>(this.queryEdges(query), edge -> {
-                return edge.label().equals(edgeLabel.name());
-            });
-        } else {
-            query.eq(HugeKeys.LABEL, edgeLabel.id());
-            return this.queryEdges(query);
+            itor = new FilterIterator<>(
+                   this.queryEdges(query),
+                   vertex -> vertex.label().equals(edgeLabel.name()));
```
Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java]: param 'vertex' -> 'edge'

Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java]: edge

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java [hunk 6]
```diff
@@ -1372,38 +1363,94 @@ public void removeEdges(EdgeLabel edgeLabel) {
...
+        Iterator<Vertex> itor;
+        // Not support label index, query all and filter by label
         if (!vertexLabel.enableLabelIndex()) {
-            return new FilterIterator<>(this.queryVertices(query), vertex -> {
-                return vertex.label().equals(vertexLabel.name());
-            });
-        } else {
-            query.eq(HugeKeys.LABEL, vertexLabel.id());
-            return this.queryVertices(query);
+            itor = new FilterIterator<>(
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java]: query.capacity(Query.NO_CAPACITY);

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java [hunk 7]
```diff
@@ -1045,31 +1045,31 @@ public void rebuildIndex(SchemaElement schema) {
...
                 assert schema.type() == HugeType.PROPERTY_KEY;
                 throw new AssertionError(String.format(
-                          "The %s can't rebuild index.", schema.type()));
+                          "The %s can't rebuild index", schema.type()));
         }
     }
 
-    public void rebuildIndex(HugeType type, Id label,
-                             Collection<Id> indexLabelIds) {
+    public void rebuildIndex(SchemaLabel label, Collection<Id> indexLabelIds) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java]: prefer move into class RebuildIndexCallable

**Pull request merged**